integration/framework/scenario.go

```diff
@@ -0,0 +1,292 @@
+package framework
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	nextScenarioID = 1
+)
+
+type Scenario struct {
+	id          int
+	networkName string
+	services    []*Service
+}
+
+func NewScenario() (*Scenario, error) {
+	s := &Scenario{
+		id:          nextScenarioID,
+		networkName: "cortex-" + strconv.Itoa(nextScenarioID),
+		services:    []*Service{},
+	}
+
+	// Setup the docker network
+	if out, err := RunCommandAndGetOutput("docker", "network", "create", s.networkName); err != nil {
+		fmt.Println(string(out))
+		return nil, errors.Wrapf(err, "create docker network '%s'", s.networkName)
+	}
+
+	nextScenarioID++
+	return s, nil
+}
+
+func (s *Scenario) Service(name string) *Service {
+	for _, service := range s.services {
+		if service.name == name {
+			return service
+		}
+	}
+
+	return nil
+}
+
+func (s *Scenario) Endpoint(name string, port int) string {
+	service := s.Service(name)
+	if service == nil {
+		return ""
+	}
+
+	return service.Endpoint(port)
+}
+
+func (s *Scenario) StartService(service *Service) error {
+	fmt.Println("Starting", service.name)
+
+	// Ensure another service with the same name doesn't exist
+	if s.Service(service.name) != nil {
+		return fmt.Errorf("Another service with the same name '%s' has already been started", service.name)
+	}
+
+	// Start the service
+	if err := service.Start(); err != nil {
+		return err
+	}
+
+	// Add to the list of services
+	s.services = append(s.services, service)
+	return nil
+}
+
+func (s *Scenario) StartConsul() error {
+	return s.StartService(NewService(
+		"consul",
+		// If you change the image tag, remember to update it in the preloading done
+		// by CircleCI too (see .circleci/config.yml).
+		"consul:0.9",
+		s.networkName,
+		[]int{},
+		nil,
+		// Run consul in "dev" mode so that the initial leader election is immediate
+		NewCommand("agent", "-server", "-client=0.0.0.0", "-dev", "-log-level=err"),
+		nil,
+	))
+}
+
+func (s *Scenario) StartMinio() error {
+	return s.StartService(NewService(
+		"minio",
+		// If you change the image tag, remember to update it in the preloading done
+		// by CircleCI too (see .circleci/config.yml).
+		"minio/minio:RELEASE.2019-12-30T05-45-39Z",
+		s.networkName,
+		[]int{9000},
+		map[string]string{
+			"MINIO_ACCESS_KEY": "cortex",
+			"MINIO_SECRET_KEY": "supersecret",
```

Comment:
It would be nice to extract these hardcoded credentials and other shared values into constants, and use them from integration/configs.go and other places.